### PR TITLE
Async redux

### PR DESCRIPTION
**Pull request by viktorgarkusha**

### CODE REVIEW

File: src/store/thunks/userThunk.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	{ dispatch: IAppDispatch; state: IRootState }
+>(
+	'users/fetchUserRole',
+	async () => {
+		const token = JSON.parse(getValue(USER_TOKEN));
+		const response = await courseApi.get('/users/me', {
+			headers: {
+				Authorization: token,
+			},
+		});
```
Comment (GolubkovN) [inline src/store/thunks/userThunk.ts]: please move it to the axios interceptor for request and use only there

File: src/store/thunks/userThunk.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	}
+);
+
+export const logout = createAsyncThunk('users/logout', async () => {
+	const token = JSON.parse(getValue(USER_TOKEN));
+	const response = await courseApi.delete('/logout', {
+		headers: {
+			Authorization: token,
+		},
+	});
```
Comment (GolubkovN) [inline src/store/thunks/userThunk.ts]: the same& move to the interceptor

File: src/types/user.ts [hunk 3]
```diff
@@ -3,4 +3,5 @@ export type UserType = {
 	email: string;
 	token: string;
 	isAuth: boolean;
+	role: string;
```
Comment (GolubkovN) [inline src/types/user.ts]: i'd recomend make it optional

File: src/store/thunks/authorsThunk.ts [hunk 4]
```diff
@@ -2,10 +2,31 @@ import { createAsyncThunk } from '@reduxjs/toolkit';
...
+	async (author: SaveAuthorType) => {
+		const token = JSON.parse(getValue(USER_TOKEN));
+		const response = await courseApi.post(
+			'/authors/add',
+			JSON.stringify(author),
+			{
+				headers: {
+					Authorization: token,
+				},
+			}
```
Comment (GolubkovN) [inline src/store/thunks/authorsThunk.ts]: interceptor

File: src/store/thunks/coursesThunk.ts [hunk 5]
```diff
@@ -1,11 +1,87 @@
...
-export const fetchAllCourses = createAsyncThunk(
-	'courses/fetchAllCourses',
-	async () => {
-		const response = await courseApi.get('/courses/all');
-		return response.data.result;
+export const fetchAllCourses = createAsyncThunk<
+	TCourses[],
+	void,
+	{ dispatch: IAppDispatch; state: IRootState }
+>('courses/fetchAllCourses', async (param, { getState }) => {
```
Comment (GolubkovN) [inline src/store/thunks/coursesThunk.ts:14]: you don't use param here^ remove

File: src/common/ProtectedRoute/AdminRoute.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+const AdminRoute = ({ redirectUrl = '/courses' }: TProtectedRoute) => {
+	const role = useAppSelector(selectUserRole);
+	useUpdateUserInStore();
+	if (role === 'admin') {
+		return <Outlet />;
+	}
+	return <Navigate to={redirectUrl} replace />;
+};
+
+export default React.memo(AdminRoute);
```
Comment (GolubkovN) [inline src/common/ProtectedRoute/AdminRoute.tsx:21]: reuse exist protectedRoute

**Review by GolubkovN — APPROVED** [review]

File: src/store/thunks/coursesThunk.ts (hunk 5 above)
Comment (viktorgarkusha) [reply src/store/thunks/coursesThunk.ts:14]: it should be of type void

File: src/common/ProtectedRoute/AdminRoute.tsx (hunk 6 above)
Comment (viktorgarkusha) [reply src/common/ProtectedRoute/AdminRoute.tsx:21]: one protected route for logged in users and another for admins only

**Pull request merged**